### DISCUSSION TITLE
Flags in compiler settings

### DISCUSSION THREAD

**Discussion by arunchawla-NOAA** [ufs-community/ufs-weather-model#1138] [Announcements]
We have three compiler options -- debug, repro and prod

prod mode settings not everything is reproducing. Repro mode is supposed to test for reproducibility with some level of optimization. In repro mode setting we have an option -fp-model consistent. This is not reproducing results but with -fp-model source the results are reproducing. 

Should we make this change ? Also should the model only be run in repro mode since that is the mode that guarantees reproducibility ?

**Comment by arunchawla-NOAA** [comment]
@junwang-noaa connected me to this document (I hope it comes through). So consistent should be a stricter standard

chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/viewer.html?pdfurl=https%3A%2F%2Fwww.intel.com%2Fcontent%2Fdam%2Fdevelop%2Fexternal%2Fus%2Fen%2Fdocuments%2Fpdf%2Ffp-consistency-121918.pdf&clen=502091

**Comment by kgerheiser** [comment]
[Intel has some good documentation on the differences in fp-model](https://www.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/floating-point-options/fp-model-fp.html)

`-fp-model source` also enables `-fp-model precise`, which may be where the difference in reproducibility is coming from. `source` is supposed to be a good tradeoff between performance and reproducibility. You can also combine `-fp-model consistent -fp-model precise`, if you want to keep `consistent`.

**Comment by DusanJovic-NOAA** [comment]
Whatever options give us run to run reproducible outputs, reproducible outputs using different MPI decomposition, and reproducible outputs after model restart, in my opinion should be a default. If that's what current 'REPRO' mode is than that should be a default, something you get without setting any specific build option.

If there are cases where users do not care about any of the above, then we can provide options that give ultimate performance but that must not be a default, it must be hidden behind an option, and it should be documented with big fat warning.

Basically, reproducibility must not be a opt-in feature. It must be an opt-out. You need to explicitly say 'I do not care about reproducibility, give me faster code', by passing -Dsomething. And, again in my opinion, that 'something' should not be 'PROD', because as far as I know production runs are required to be reproducible.

**Comment by arunchawla-NOAA** [comment]
we should do a timing run on prod vs repro mode to see what the differences are

**Comment by arunchawla-NOAA** [comment]
Also based on the experiments being done by @RatkoVasic-NOAA (see https://github.com/ufs-community/ufs-weather-model/issues/649#issuecomment-1076700839 ) not all results are being fixed by this

**Comment by climbfuji** [comment]
We did that 2-3 years back when REPRO was introduced first (for b4b reproducibility between IPD and CCPP, testing MPI decomp reproducibility for regional runs wasn’t a question back then). At that time it was about 10% give or take, mainly due to the AVX2 flags (the others were negligible).  But this was NEMSfv3gfs atm only back then, and nearly all the time was spent in the dycore for these ATM only runs.

**Comment by arunchawla-NOAA** [comment]
Turns out GOCART is reproducible in REPRO mode and we do not have to switch from -fp-model consistent to -fp-model source, which is good because it was creating confusion as to why this was a problem. However it does raise a question on which of the flags in the PROD mode are really needed for speeding up the codes. Dom's suggestion is that only the AVX flag makes a big difference

**Comment by arunchawla-NOAA** [comment]
To recap our opsreqtest does the following

1. compile with debug and run to completion
2. compile with single and double precision and run to completion
3. compile and test reproducibility with following options -- change threading, change mpi count, change domain decomposition, and restart from a restart file

We want our codes to satisfy these tests -- in conversation with NCO they have indicated they only want those compiler flags that satisfy the above tests. 

So we need to decide what compiler options give us that

**Comment by arunchawla-NOAA** [comment]
From science reproducibility and testing point of view we should only focus on our testing on 2 sets of compiler settings -- one for debug and one standard that will expect reproducibility with the code

Jun pointed to this presentation which provides a discussion 

[https://www.ecmwf.int/sites/default/files/elibrary/2012/14026-optimisation-weather-[…]ications-power-and-x86-architectures-focus-reproducibility.pdf](https://www.ecmwf.int/sites/default/files/elibrary/2012/14026-optimisation-weather-applications-power-and-x86-architectures-focus-reproducibility.pdf)

The performance hit is about 10-20%

**Comment by arunchawla-NOAA** [comment]
So what do we consider "safe compiler flags?"

Right now in repro mode we use

-g -traceback -fpp -fno-alias -auto -safe-cray-ptr -ftz -assume byterecl -nowarn -sox -align array64byte -qno-opt-dynamic-align -O2 -debug minimal -fp-model consistent -qoverride-limits

Plus small changes for 32 bit , openmp etc

**Comment by arunchawla-NOAA** [comment]
The additional flags that we use for performance are

-qopt-prefetch=3 -no-prec-div -no-prec-sqrt

plus the different AVX recipes

These are what we use for performance. Perhaps we should just drop them in our testing and in operations ?

**Comment by climbfuji** [comment]
>  -no-prec-div -no-prec-sqrt
> 
is only used in double precision (64bit), not in single precision (32bit). To my experience, they don’t do much to the performance.

Agressive prefetching can be a problem, and I didn’t see a huge impact of -qopt-prefetch=3 on the performance either.

The only ones that really matter for performance are AVX2 and the -fp-model choice, in my experience. Whether you need to remove SIMD (AVX2) entirely or substitute with core-avx-i or something like that needs to be seen.

Note that the PDF that you are pointing to is from 2012 and while generally valid, a lot happened on the compiler side since then (e.g. -fp-model consistent didn’t exist back then, therefore they used precise). Note also that they used -O3, which is more agressive than our -O2 - not sure how much that helps with performance and what it does with reproducibility in our case.

**Comment by junwang-noaa** [comment]
I want to point out that the compile options such as -ftz or fp-model consistent do not guarantee the reproducibility. So in general, what is the approach for us to get reproducibility ( only using guaranteed reproducibility compile options or test-and-see approach) for ufs-weather-model?

**Reply by arunchawla-NOAA** [reply]
I am guessing yes, as these area is always changing. But happy to hear other suggestions. One thing we need to be careful about is not choosing the tests that match our codes as opposed to cleaning our codes

**Comment by arunchawla-NOAA** [comment]
Intel has some documentation as well

https://www.intel.com/content/dam/develop/external/us/en/documents/pdf/fp-consistency-121918.pdf

**Comment by JamesAbeles-NOAA** [comment]
OK, we need two things as Arun's email pointed out
1 - consistent, reproducible answers
2 - a configuration to meet the operational runtime requirement
We should focus on 1 first to determine which flags will give reproducible results for the criteria described in earlier comments. Then we can see what configuration can meet the runtime. If we find the runtime is too slow meeting (1) above, then we need to play with flags to see if we can get better runtime while still meeting the requirement

**Comment by RatkoVasic-NOAA** [comment]
For now none of options reproduce regional hi-res case (LAM parallel). Only low-resolution case reproduces:
https://docs.google.com/spreadsheets/d/1RUK063Ml3ES2-wXhk-WymJpf4K1soLmOSpmf7NvTjr8/edit?usp=sharing

**Reply by arunchawla-NOAA** [reply]
Ratko sorry I did not reply to this. Once we have chosen the final flags for release mode we will want reproducibility so this issue will need to get addressed for HAFS as well as regional

**Comment by edwardhartnett** [comment]
Better component-level testing would reduce the importance of these whole-system tests.

**Comment by arunchawla-NOAA** [comment]
Hi, we would like to finalize this discussion this week. We would like to replace PROD and REPRO with just one set of flags, for lack of better word lets call it STD. 

So far we have the the REPRO flags to which we can add AVX. Do we need anymore ?

**Reply by DusanJovic-NOAA** [reply]
Why not just 'Release' and 'Debug' which are standard cmake build types, as @aerorahul suggested.

**Reply by arunchawla-NOAA** [reply]
Release works, whatever is the standard nomenclature. No need to invent new ones

**Comment by JamesAbeles-NOAA** [comment]
I agree with Dusan.

**Comment by climbfuji** [comment]
Me to. Then we don’t need to translate the PROD and DEBUG flags into Release and Debug for CCPP anymore. Just adjust the CCPP Release type and remove the Bitforbit type.

**Comment by arunchawla-NOAA** [comment]
Trying to put a pin in this discussion. So we are agreeing on

1. remove the repro mode
2. replace the name PROD with RELEASE
3. Get rid of -qopt-prefetch=3 -no-prec-div -no-prec-sqrt

Does this sound right ?

**Reply by JessicaMeixner-NOAA** [reply]
I think so.  I think we also have to use -fp-model consistent instead of -fp-model source

**Reply by arunchawla-NOAA** [reply]
that is correct. Huh, we use -fp-model consistent in REPRO but not in PROD mode.

**Comment by RatkoVasic-NOAA** [comment]
@BenjaminBlake-NOAA just did test on LAM parallel using PROD vs. REPRO. PROD was about 12% faster.
Differences in compilation:
```
PROD:  -fp-model source     -qopt-prefetch=3 -march=core-avx2
REPRO: -fp-model consistent
```
Not sure how each option contribute to the speedup/slowdown, but if necessary, we could do the tests.

**Comment by JamesAbeles-NOAA** [comment]
I do not think -qopt-prefetch=3 does anything except to add compile time. I
suspect the main thing will be use of avx2.

On Tue, Apr 5, 2022 at 1:05 PM RatkoVasic-NOAA ***@***.***>
wrote:

> @BenjaminBlake-NOAA <https://github.com/BenjaminBlake-NOAA> just did test
> on LAM parallel using PROD vs. REPRO. PROD was about 12% faster.
> Differences in compilation:
>
> PROD:  -fp-model source     -qopt-prefetch=3 -march=core-avx2
> REPRO: -fp-model consistent
>
> Not sure how each option contribute to the speedup/slowdown, but if
> necessary, we could do the tests.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ufs-community/ufs-weather-model/discussions/1138#discussioncomment-2509853>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALYJU7ZDHPPCUEGELWWANI3VDRXD7ANCNFSM5ROXQSUA>
> .
> You are receiving this because you commented.Message ID:
> <ufs-community/ufs-weather-model/repo-discussions/1138/comments/2509853@
> github.com>
>


-- 
Jim Abeles
*IMSG* at NOAA/NWS/NCEP/EMC
301 879-3283

**Reply by climbfuji** [reply]
Yes, this confirms our earlier testing (2 years ago) that avx2 speeds the code up by 10-15%.

**Comment by arunchawla-NOAA** [comment]
Based on discussions this is the final compile settings that we will use in our tests

1. remove the repro mode
2. replace the name PROD with RELEASE
3. update the flags for RELEASE to 
     a. Get rid of -qopt-prefetch=3 -no-prec-div -no-prec-sqrt options
     b. replace -fp-model source with -fp-model consistent

**Comment by DusanJovic-NOAA** [comment]
See PR #1171
